sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/WindowedSubscriber.java

```diff
@@ -0,0 +1,929 @@
+        final EnqueueResult<T> r = enqueueRequestImpl(windowSize, windowTimeout, windowSink);
+        return r.getWindowIterable();
+    }
+
+    // Opened in Package-private scope for testing purposes, any library need will use 'enqueueRequest' API.
+    EnqueueResult<T> enqueueRequestImpl(int windowSize, Duration windowTimeout, Sinks.Many<T> windowSink) {
+        if (windowSize < 1) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'windowSize' must be strictly positive."));
+        }
+        if (Objects.isNull(windowTimeout)) {
+            throw logger.logExceptionAsError(new NullPointerException("'windowTimeout' cannot be null."));
+        }
+        if (windowTimeout.isNegative() || windowTimeout.isZero()) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("'windowTimeout' period must be strictly positive."));
+        }
+        if (Objects.isNull(windowSink)) {
+            throw logger.logExceptionAsError(new NullPointerException("'windowSink' cannot be null."));
+        }
+
+        final long workId = idGenerator.getAndIncrement();
+        final WindowWork<T> w = new WindowWork<T>(this, workId, windowSize, windowTimeout, windowSink);
+        if (isDoneOrCancelled()) {
+            w.terminate(WorkTerminalState.PARENT_TERMINATED);
+            return new EnqueueResult<T>(w, w.windowFlux(false));
+        }
+
+        workQueue.add(w);
+        drain();
+        return new EnqueueResult<T>(w, w.windowFlux(true));
+    }
+
+    @Override
+    protected void hookOnSubscribe(Subscription s) {
+        if (Operators.setOnce(S, this, s)) {
+            drain();
+        }
+    }
+
+    @Override
+    protected void hookOnNext(T item) {
+        if (done) {
+            Operators.onNextDropped(item, super.currentContext());
+            return;
+        }
+
+        if (s == Operators.cancelledSubscription()) {
+            Operators.onDiscard(item, super.currentContext());
+            return;
+        }
+
+        if (queue.offer(item)) {
+            drain();
+        } else {
+            // The item 'queue' is 'unbounded' so this code block won't be hit, still follow guideline.
+            final RuntimeException e = Exceptions.failWithOverflow(Exceptions.BACKPRESSURE_ERROR_QUEUE_FULL);
+            Operators.onOperatorError(this, e, super.currentContext());
+            Operators.onDiscard(item, super.currentContext());
+            if (ERROR.compareAndSet(this, null, e)) {
+                done = true;
+                drain();
+            } else {
+                done = true;
+                Operators.onErrorDropped(e, super.currentContext());
+            }
+        }
+    }
+
+    @Override
+    protected void hookOnError(Throwable t) {
+        if (done) {
+            Operators.onErrorDropped(t, super.currentContext());
+            return;
+        }
+
+        final RuntimeException e = new RuntimeException(String.format(terminatedMessage, "upstream-error"), t);
```

Comment:
do we want to distinguish retriable and non-retriable errors somehow?

Reply:
Hi @lmolkova, In V2, the upstream of the `WindowedSubscriber` looks like below -

![MessageChain](https://github.com/Azure/azure-sdk-for-java/assets/1471612/ff1b7221-495a-45ca-a3e0-c55e78cdbd45)

So, long before this (upstream) terminal error arrives in `WindowedSubscriber`, the upstream layers would have already done inspections and the possible retries, without `WindowedSubscriber` realizing it.

At this point I was thinking we propagate the error (retry exhaust, unknown error) by wrapping in RuntimeException without inspecting locally.

The owner of ` WindowedSubscriber` responsible to decides what to do.  Today, the owner (client) returns an error to the application requesting to re-create the client. (This could be improved, those ideas you mentioned in our offline team-chat).

Reply:
thank you for the explanation!